game/procurement.py

```diff
@@ -116,52 +128,236 @@ def repair_runways(self, budget: float) -> float:
+                if (
+                    cp.base.total_ground_units
+                    > cp_with_fewest_units.base.total_ground_units
+                ):
+                    cp_with_fewest_units = cp
+
+            groundunits_priority_list: List[
+                Tuple[List, float]
+            ] = self.calculate_vehicle_investment_ratio(cp_with_fewest_units)
+            # if no priority is found, buy a tank
+            most_needed_type = TYPE_TANKS
+            if groundunits_priority_list:
+                lowest_ratio: float = 99
+                for gu_priority in groundunits_priority_list:
+                    if gu_priority[1] < lowest_ratio:
+                        lowest_ratio = gu_priority[1]
+                        most_needed_type = gu_priority[0]
+
+            unit = self.buy_ground_unit(
+                ground_unit_budget, most_needed_type, cp_with_fewest_units
+            )
             if unit is None:
-                # Can't afford any more units.
                 break
 
-            budget -= db.PRICES[unit]
-            cp.pending_unit_deliveries.order({unit: 1})
+            unit_price = db.PRICES[unit]
+            ground_unit_budget -= unit_price
+            cp_with_fewest_units.pending_unit_deliveries.order({unit: 1})
 
-        return budget
+        return ground_unit_budget
+
+    def buy_ground_unit(
+        self, budget: float, vehicle_type: List[Type[VehicleType]], cp: ControlPoint
+    ) -> Optional[Type[VehicleType]]:
+        unit = self.random_affordable_ground_unit(budget, vehicle_type)
+
+        if unit is None:
+            # Can't afford any more units.
+            return None
+
+        return unit
+
+    def calculate_vehicle_investment_ratio(
+        self, cp: ControlPoint
+    ) -> List[Tuple[List, float]]:
+        prioritylist: List[Tuple[List, float]] = []
+
+        tank_costs = self.calculate_investment_of_vehicleType(cp, TYPE_TANKS)
+        atgm_costs = self.calculate_investment_of_vehicleType(cp, TYPE_ATGM)
+        ifv_costs = self.calculate_investment_of_vehicleType(cp, TYPE_IFV)
+        apc_costs = self.calculate_investment_of_vehicleType(cp, TYPE_APC)
+        artillery_costs = self.calculate_investment_of_vehicleType(cp, TYPE_ARTILLERY)
+        shorad_costs = self.calculate_investment_of_vehicleType(cp, TYPE_SHORAD)
+
+        costs_of_all_vehicles = (
+            tank_costs
+            + atgm_costs
+            + ifv_costs
+            + apc_costs
+            + artillery_costs
+            + shorad_costs
+        )
+
+        faction_has_tank_access = False
+        faction_has_atgm_access = False
+        faction_has_ifv_access = False
+        faction_has_apc_access = False
+        faction_has_artillery_access = False
+        faction_has_shorad_access = False
+
+        for unit in self.faction.frontline_units:
+            if faction_has_tank_access is False and unit in TYPE_TANKS:
+                faction_has_tank_access = True
```

Comment:
Make this a [`cached_property`](https://docs.python.org/3/library/functools.html#functools.cached_property) of the faction?